pkg/awsutils/awsutils.go

```diff
@@ -674,17 +689,50 @@ func (cache *EC2InstanceMetadataCache) AllocIPAddress(eniID string) error {
 	return nil
 }
 
+// GetENIipLimit return IP address limit per ENI based on EC2 instance type
+func (cache *EC2InstanceMetadataCache) GetENIipLimit() (int64, error) {
+	ipLimit, ok := InstanceIPsAvailable[cache.instanceType]
+	if !ok {
+		log.Errorf("Failed to get eni IP limit due to unknown instance type %s", cache.instanceType)
+		return 0, errors.New(UnknownInstanceType)
+	}
+
+	return ipLimit - 1, nil
+}
+
 // AllocAllIPAddress allocates all IP addresses available on eni
 func (cache *EC2InstanceMetadataCache) AllocAllIPAddress(eniID string) error {
 	log.Infof("Trying to allocate all available ip addresses on eni: %s", eniID)
 
-	input := &ec2.AssignPrivateIpAddressesInput{
-		NetworkInterfaceId:             aws.String(eniID),
-		SecondaryPrivateIpAddressCount: aws.Int64(1),
-	}
+	ipLimit, err := cache.GetENIipLimit()
+
+	if err != nil {
+		// for unknown instance type, will allocate one ip address at a time
+		ipLimit = 1
+
+		input := &ec2.AssignPrivateIpAddressesInput{
+			NetworkInterfaceId:             aws.String(eniID),
+			SecondaryPrivateIpAddressCount: aws.Int64(ipLimit),
+		}
+
+		for {
+			// until error
+			_, err := cache.ec2SVC.AssignPrivateIpAddresses(input)
+			if err != nil {
+				if containsPrivateIPAddressLimitExceededError(err) {
```

Comment:
I don't like the name `containsPrivateIPAddressLimitExceededError`.  Yes, I understand the you have to unpack the error to see if it's the one you're testing, but that's an implementation detail.  Prefer the plainer`isPrivateIPAddressLimitExceededError` for this test function.